### PR TITLE
[C] Flex Layout (flex c engine)

### PR DESCRIPTION
**Pull request by StephaneDelcroix**
### Description of Change ###

Add a new FlexLayout to Xamarin.Forms.

#### Credits

The current implementation is a wrapper around @lrz's https://github.com/xamarin/flex.

Thanks to @rmarinho for the initial proof of concept, the almost-extensive test suite and the packaging.

#### TODO
 - [ ] Test the packaging, install and use on sim and devices from nuget
 - [ ] Test the previewer (iOS should load the framework)
 - [ ] Add TypeConverter on Basis for Xaml usage. TypeConverters on all enums for CSS usage
 - [ ] Add CSS mappings

### Bugs Fixed ###

/

### API Changes ###

Added:
```
public class Xamarin.Forms.FlexLayout : Layout<View>
{
	//BP, applies to the layout
	public static readonly BindableProperty DirectionProperty;
	public static readonly BindableProperty JustifyContentProperty;
	public static readonly BindableProperty AlignContentProperty;
	public static readonly BindableProperty AlignItemsProperty;
	public static readonly BindableProperty PositionProperty;
	public static readonly BindableProperty WrapProperty;

	//Attached ones, applies to children
	public static readonly BindableProperty OrderProperty;
	public static readonly BindableProperty GrowProperty;
	public static readonly BindableProperty ShrinkProperty;
	public static readonly BindableProperty AlignSelfProperty;
	public static readonly BindableProperty BasisProperty;

	//property accessors
	public FlexDirection Direction { get; set; }
	public FlexJustify JustifyContent { get; set; }
	public FlexAlignContent AlignContent { get; set; }
	public FlexAlignItems AlignItems { get; set; }
	public FlexPosition Position { get; set; }
	public FlexWrap Wrap { get; set; }

	public static int GetOrder(BindableObject bindable);
	public static void SetOrder(BindableObject bindable, int value);
	public static float GetGrow(BindableObject bindable);
	public static void SetGrow(BindableObject bindable, float value);
	public static float GetShrink(BindableObject bindable);
	public static void SetShrink(BindableObject bindable, float value);
	public static FlexAlignSelf GetAlignSelf(BindableObject bindable);
	public static void SetAlignSelf(BindableObject bindable, FlexAlignSelf value);
	public static FlexBasis GetBasis(BindableObject bindable);
	public static void SetBasis(BindableObject bindable, FlexBasis value);
}

public enum Xamarin.Forms.FlexJustify
{
	FlexStart,
	Center,
	FlexEnd,
	SpaceBetween,
	SpaceAround,
	SpaceEvenly,
}

public enum Xamarin.Forms.FlexPosition
{
	Relative,
	Absolute,
}

public enum Xamarin.Forms.FlexDirection
{
	Column,
	ColumnReverse,
	Row,
	RowReverse,
}

public enum Xamarin.Forms.FlexAlignContent
{
	Stretch,
	Center,
	FlexStart,
	FlexEnd,
	SpaceBetween,
	SpaceAround,
}

public enum Xamarin.Forms.FlexAlignItems
{
	Stretch,
	Center,
	FlexStart,
	FlexEnd,
}

public enum Xamarin.Forms.FlexAlignSelf
{
	Auto,
	Stretch,
	Center,
	FlexStart,
	FlexEnd,
}

public enum Xamarin.Forms.FlexWrap
{
	NoWrap,
	Wrap,
	Reverse,
}

public struct Xamarin.Forms.FlexBasis
{
	public static FlexBasis Auto;
	public float Length { get; }
	public FlexBasis(float length);
	public static implicit operator FlexBasis(float length);
}
```

### Behavioral Changes ###

/

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of master at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Core/Flex.cs [hunk 1]
```diff
@@ -0,0 +1,340 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/Flex.cs]: the scores directly from Xamarin.Flex, but changed to internal

File: Xamarin.Forms.Core/FlexBase.cs [hunk 2]
```diff
@@ -0,0 +1,426 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/FlexBase.cs]: same for this

File: Xamarin.Forms.Core/FlexEnums.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
+﻿using System;
+namespace Xamarin.Forms
+{
+	//TODO require a typeConverter for css values
+	public enum FlexJustify
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/FlexEnums.cs]: As Xamarin.Flex enums sometimes have values not supported for some properties, we redefine them

File: Xamarin.Forms.Core/FlexEnums.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+	//TODO require a typeConverter for css values
+	public enum FlexWrap
+	{
+		NoWrap = Flex.Wrap.NoWrap,
+		Wrap = Flex.Wrap.Wrap,
+		Reverse = Flex.Wrap.WrapReverse,
+	}
+
+	//TODO require a typeConverter
+	public struct FlexBasis
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/FlexEnums.cs]: Xamarin.Flex uses `float.NaN` as `Auto`

File: Xamarin.Forms.Core/FlexLayout.cs [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+using System.ComponentModel;
+using System.Runtime.CompilerServices;
+using Xamarin;
+using Xamarin.Forms.Internals;
+
+namespace Xamarin.Forms
+{
+	public class FlexLayout : Layout<View>
+	{
+		public static readonly BindableProperty DirectionProperty =
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/FlexLayout.cs]: boilerplate from here....

File: Xamarin.Forms.Core/FlexLayout.cs [hunk 6]
```diff
@@ -0,0 +1,452 @@
...
+		}
+
+		static void OnWrapPropertyChanged(BindableObject bindable, object oldValue, object newValue)
+		{
+			var flexLayout = bindable as FlexLayout;
+			if (flexLayout._root == null)
+				return;
+			flexLayout._root.Wrap = (Flex.Wrap)(FlexWrap)newValue;
+			flexLayout.InvalidateLayout();
+		}
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/FlexLayout.cs]: ...to there

File: Xamarin.Forms.Core/FlexLayout.cs [hunk 7]
```diff
@@ -0,0 +1,452 @@
...
+				return;
+			var item = GetFlexItem(view);
+			_root.Remove(item);
+			item.Dispose();
+			view.ClearValue(FlexItemProperty);
+		}
+
+		void OnChildPropertyChanged(object sender, PropertyChangedEventArgs e)
+		{
+			if (   e.PropertyName == WidthRequestProperty.PropertyName
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/FlexLayout.cs]: the 2 cases can probably be merged

File: Xamarin.Forms.Core.UnitTests/FlexLayoutAlignContentTests.cs [hunk 8]
```diff
@@ -0,0 +1,753 @@
...
+			Assert.AreEqual(100f, view3.Height);
+
+			Assert.AreEqual(100f, view4.X);
+			Assert.AreEqual(0f, view4.Y);
+			Assert.AreEqual(50f, view4.Width);
+			Assert.AreEqual(100f, view4.Height);
+		}
+
+		[Test]
+		[Ignore("")]
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core.UnitTests/FlexLayoutAlignContentTests.cs]: this is the last issue that needs to be fixed in Xamarin.Flex.

**Comment by bugadani** [comment]
Cool, but I'm wondering if native code is the best decision here. Is/will this be supported on Tizen, macOS and GTK?

**Comment by rmarinho** [comment]
@bugadani it will work on all platforms. we will have flex bindings for all.

**Comment by bentmar** [comment]
Will we get any performance boosts using flex? :)

**Comment by StephaneDelcroix** [comment]
@bentmar layout complexity is O(n)

**Comment by StephaneDelcroix** [comment]
see #1759

**Comment by bentmar** [comment]
so this will bring performance better that say grid or stacklayout?

**Pull request closed without merging**